packages/headless/src/controllers/controller/headless-controller.ts

```diff
@@ -35,12 +37,25 @@ export function buildController<T extends object>(
   return {
     subscribe(listener: () => void) {
       listener();
-      prevState = JSON.stringify(this.state);
-      return engine.subscribe(() => {
-        if (hasStateChanged(this.state)) {
-          listener();
+      const symbol = Symbol();
+      let unsubscribe: () => void;
+
+      if (hasNoListeners()) {
+        prevState = JSON.stringify(this.state);
+        unsubscribe = engine.subscribe(() => {
+          if (hasStateChanged(this.state)) {
+            listeners.forEach((listener) => listener());
+          }
+        });
```

Comment:
Seeing this code with fresh eyes this morning,

I'm wondering if using closure can avoid using a "global" `prevState` and `listeners` map, moving from a 1:many to a 1:1 relationship.

The `hasStateChanged` will need to adjust to not write, and can return the stringified json for a small performance win.

```
const statefulListener = (() => {
    let prevState = JSON.stringify(this.state);
    return () => {
        const {yes, stringifiedState} = hasStateChanged(this.state);
       if (yes) {
          prevState = stringifiedState;
          listener();
       }
    }
})()

return engine.subscribe(statefulListener)
```

Another benefit is there would be no need to create a custom `unsubscribe` to clean up global state,

Reply:
Then we are back to the first PR I made, where we do N comparison for N listener, no ?

Reply:
Indeed, I overlooked that these subscribes are called on a single component instance,
Will approve the current solution,